src/WebJobs.Extensions.DurableTask/Auth/AppAuthenticationCredentialsFactory.cs

```diff
@@ -0,0 +1,220 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Services.AppAuthentication;
+using Microsoft.Extensions.Logging;
+using Microsoft.WindowsAzure.Storage.Auth;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Auth
+{
+    // TODO: Replace with Azure.Identity
+    internal sealed class AppAuthenticationCredentialsFactory : IStorageCredentialsFactory, IDisposable
+    {
+        private const string LoggerName = "Host.Triggers.DurableTask.Auth";
+
+        private readonly Func<string, AzureServiceTokenProvider> tokenProviderFactory;
+        private readonly ILogger logger;
+        private readonly AsyncLock cacheLock = new AsyncLock();
+        private readonly Dictionary<string, TokenCredential> cache = new Dictionary<string, TokenCredential>();
+        private volatile bool disposed;
+
+        public AppAuthenticationCredentialsFactory(ILoggerFactory loggerFactory)
+            : this(s => new AzureServiceTokenProvider(s), loggerFactory)
+        { }
+
+        internal AppAuthenticationCredentialsFactory(Func<string, AzureServiceTokenProvider> tokenProviderFactory, ILoggerFactory loggerFactory)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            this.tokenProviderFactory = tokenProviderFactory ?? throw new ArgumentNullException(nameof(tokenProviderFactory));
+            this.logger = loggerFactory.CreateLogger(LoggerName);
+        }
+
+        internal event Action<TokenRenewalState> Renewing;
+
+        internal event Action<NewTokenAndFrequency> Renewed;
+
+        internal event Action<NewTokenAndFrequency, AzureServiceTokenProviderException> RenewalFailed;
+
+        public Task<StorageCredentials> CreateAsync(AzureIdentityOptions options, CancellationToken cancellationToken = default) =>
+            this.CreateAsync(options, TimeSpan.FromMinutes(5d), TimeSpan.FromSeconds(30d), cancellationToken);
```

Comment:
Ah, I shouldn't have included the `d` suffix because I too have been reading it as "5 days" lol. By default we are renewing tokens 5 minutes before they expire and we retry renewal after a simple 30 seconds.

Reply:
Oh! I totally misread them as 5 days and 30 days (this is how you specify a timespan in days in Kusto)!